### ISSUE TITLE
supervisorctl "always" exits with zero

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach-prod#59]
That's just really bad.

```bash
ubuntu@ip-172-31-55-72:~$ supervisorctl asd
*** Unknown syntax: asd
ubuntu@ip-172-31-55-72:~$ echo $?
0
ubuntu@ip-172-31-55-72:~$ supervisorctl start
unix:///var/run/supervisor.sock no such file
ubuntu@ip-172-31-55-72:~$ echo $?
0
ubuntu@ip-172-31-55-72:~$ supervisorctl -c supervisor.conf start
Error: start requires a process name
start <name>		Start a process
start <gname>:*		Start all processes in a group
start <name> <name>	Start multiple processes or groups
start all		Start all processes
ubuntu@ip-172-31-55-72:~$ echo $?
0
ubuntu@ip-172-31-55-72:~$ supervisorctl -c supervisor.conf start foobar
foobar: ERROR (no such process)
ubuntu@ip-172-31-55-72:~$ echo $?
0
```

**Comment by mnaberez** [comment]
> That's just really bad. The issue I just pinged this thread from shows this has been open for a while, and while a contributor did substantial work on it, it was never merged (nor did it seem to get nursed a lot).

You write as if the maintainers have abandoned it.  The status is that it's a current, open pull request where one of the committers (me) has already agreed that we are going to merge it and it will be included in the next major version.  As far as "nor did it seem to get nursed a lot", there were actually several iterations of patches that had various discussions, so I don't think our issue tracker supports that comment.  I get that you are frustrated that Supervisor doesn't support this yet, but we have agreed to fix it, and your comments seem to be written in a purposely negative way.

**Comment by tschottdorf** [comment]
Sorry @mnaberez, didn't actually see the open PR - I scrolled right past that https://github.com/Supervisor/supervisor/issues/641 on the thread. In any case, didn't mean to come across as harsh (looking at the comment now that's certainly the case though). Thanks for the comment.

**Comment by mnaberez** [comment]
No problem.  For background, the reason we did not merge it into Supervisor 3.x is that I considered it a BC break.  It is stupid that `supervisorctl` does not have exit codes for errors, but it's been like this for many years now, and scripts controlling supervisorctl expect the current behavior.  I didn't want to break it on anybody in a minor release.  We just released what is very possibly the last 3.x version a few days ago.  We are going to merge this for 4.0.  We don't have a release date yet but hopefully it's not too far out.

**Comment by tschottdorf** [comment]
Yeah, definitely a BC break. I'm just really surprised that in 4.5 years there's never been a change - maybe I'm making this sound easier than it is, but maintaining compatibility sounds fairly doable with a configuration setting whose default switches to `yes` in `4.0` (and justifiable through the interest in the feature, which only picks up in 2014 though, oddly enough). Happy to hear it's on for the next-ish version.

**Comment by mnaberez** [comment]
I agree it's been way too long.  The committers should have done something about it a long time ago but we didn't (not intentional, we had other issues, I guess).  The only serious patch from a contributor to fix this came recently.  Your point about a config file setting is fair but hopefully 4.0 isn't too far out at this point where that will be necessary.  My opinion is that the current behavior where supervisorctl never sets exit codes is not sane.  4.0 should just always set exit codes.

**Comment by tamird** [comment]
The upstream issue has been closed, but a supervisor version containing the fix has not yet been released. My understanding is that it's blocked on everything in https://github.com/Supervisor/supervisor/labels/python%203.

Pretty absurd, but here we are.
